docs/pages/access-controls/login-rules/reference.mdx

```diff
@@ -0,0 +1,678 @@
+---
+title: Login Rules Reference (Preview)
+description: Reference documentation for Login Rules
+---
+
+<Admonition type="warning" title="Preview">
+  Login Rules are currently in Preview mode.
+</Admonition>
+
+This page provides details on the expression language that powers Login Rules.
+To learn how to add the first login rule to your cluster, checkout out the
+[Login Rules Guide](./guide/).
+
+## YAML specification
+
+(!docs/pages/includes/login-rule-spec.mdx!)
+
+## `traits_map` vs `traits_expression`
+
+Every login rule spec must contain either the `traits_map` field or a
+`traits_expression` field.
+
+They both serve the same purpose of transforming user traits.
+The difference lies only in the syntax you prefer for your use case, since you can write
+every `traits_map` as an equivalent `traits_expression`.
+
+### `traits_map`
+
+Here is an example Login Rule using a `traits_map` which implements the
+following rules:
+1. Every user with the `groups: devs` trait should receive an extra trait
+   `access: [staging]`.
+1. Every user with the `groups: admins` trait should receive an extra trait
+   `access: [staging, prod]`.
+1. Every user should receive a `logins` trait with the value of their incoming
+   `username` trait converted to lowercase.
+1. All traits other than `groups`, `logins`, and `access` should be filtered
+   out.
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: my_expression_rule
+spec:
+  priority: 0
+
+  traits_map:
+    # the groups trait will be copied unmodified. Do the same for all other
+    # traits which should not be changed, any traits omitted here will *not* be
+    # set for your users and will *not* be used for role mapping.
+    groups:
+      - external["groups"]
+
+    # the logins traits will be set to the username trait converted to
+    # lowercase.
+    logins:
+      - 'strings.lower(external.username)'
+   
+    # the access trait is determined conditionally based on the incoming groups trait.
+    access:
+      - 'ifelse(external.groups.contains("devs"), set("staging"), set())'
+      - 'ifelse(external.groups.contains("admins"), set("staging", "prod"), set())'
+```
+
+### `traits_expression`
+
+Here is an example login rule using the `traits_expression` field implementing
+the same rules as the above example:
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: my_expression_rule
+spec:
+  priority: 0
+
+  traits_expression: |
```

Comment:
Out of curiousity, if you create this rule and then `tctl get` it, does the YAML unmarshaler ditch the whitespace and put the entire expression on one line?

Reply:
It actually comes out much nicer than I expected
```
$ tctl get login_rules/my_expression_rule
kind: login_rule
metadata:
  name: my_expression_rule
spec:
  priority: 0
  traits_expression: |
    dict(
      pair("groups", external.groups),
      pair("logins", strings.lower(external.username)),
      pair("access",
        choose(
          option(external.groups.contains("devs"), set("staging")),
          option(external.groups.contains("admins"), set("staging", "prod")),
          option(true, set()),
        ),
      ),
    )
version: v1
```